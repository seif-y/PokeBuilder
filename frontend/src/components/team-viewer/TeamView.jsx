import Body from "./util/style-components/Body";
import Headline from "./util/style-components/Headline";
import UpvotableContent from "./util/upvotes/UpvotableContent";
import { useContext, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import styles from "./TeamView.module.css";
import { PokemonDataContext } from "../../pokeapi/PokemonDataContextProvider";
import BlackHeadingTag from "../global/BlackHeadingTag";

function Details({ creatorUsername, teamName, description }) {
    const CHAR_LIMIT = 400;
    return (
        <>
            <Headline classes={`flex ${styles.title}`}>
                <BlackHeadingTag text={teamName} size={300} leftAlign />
                <span className={styles.creator}>{`by ${creatorUsername}`}</span>
            </Headline>
            <Body>{description < CHAR_LIMIT ? description : `${description.slice(0, CHAR_LIMIT)}...`}</Body>
        </>
    );
}

function PartySprites({ party }) {
    return (
        <div>
            {party.map(({ pokemonID, name, sprite }) => (
                <img
                    key={pokemonID}
                    className={styles.partySprite}
                    src={sprite}
                    title={`${name} #${pokemonID}`}
                    alt={`#${pokemonID}`}
                />
            ))}
        </div>
    );
}

export default function TeamView({
    teamData: { _id: teamID, creatorUsername, teamName, description, party },
    onVote,
    isUpvoted,
    upvotes,
}) {
    const allPokemonViews = useContext(PokemonDataContext);
    const [formattedParty, setFormattedParty] = useState([]);
    useEffect(() => {
        if (allPokemonViews) {
            setFormattedParty(
                party.map((pokemon) => {
                    const pokemonView = allPokemonViews.find((element) => element.id === pokemon.pokemonID);
                    return Object.assign({}, pokemon, pokemonView);
                })
            );
        }
    }, [party, allPokemonViews]);

    return (
        <UpvotableContent
            classes={styles.clickable}
            isUpvoted={isUpvoted}
            onVote={(isUpvoted) => onVote(isUpvoted, teamID)}
            upvotes={upvotes}
        >
            <Link className={styles.link} to={`/teams/${teamID}`}>
                <Details creatorUsername={creatorUsername} teamName={teamName} description={description} />
                <PartySprites party={formattedParty} />
            </Link>
        </UpvotableContent>
    );
}
