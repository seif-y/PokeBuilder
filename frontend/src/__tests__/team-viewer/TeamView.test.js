import { StaticRouter } from "react-router";
import TeamView from "../../components/team-viewer/TeamView";
import renderer from "react-test-renderer";

const team = {
    _id: "304u012h5k436h",
    creatorUsername: "TestUser",
    teamName: "Test Team",
    description: "This is not a real team. You have been deceived",
    party: [3, 6, 9, 12, 15, 18].map((id, ind) => {
        return { pokemonID: id, notes: `Team member #${ind}` };
    }),
};

test("TeamView snapshot for test pokemon matches", () => {
    const tree = renderer
        .create(
            <StaticRouter>
                <TeamView teamData={team} />
            </StaticRouter>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
