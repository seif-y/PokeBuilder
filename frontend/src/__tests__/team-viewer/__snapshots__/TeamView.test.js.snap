// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamView snapshot for test pokemon matches 1`] = `
<div
  className="shadowedBox flex"
>
  <div
    className="flex upvoteBox"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root  arrow clickable MuiSvgIcon-fontSizeLarge"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 8V4l8 8-8 8v-4H4V8z"
      />
    </svg>
    0
  </div>
  <div
    className="content clickable"
  >
    <a
      className="link"
      href="/teams/304u012h5k436h"
      onClick={[Function]}
    >
      <div
        className="headline flex title"
      >
        <div
          className="header"
          style={
            Object {
              "paddingLeft": 10,
              "textAlign": "left",
              "width": 300,
            }
          }
        >
          Test Team
        </div>
        <span
          className="creator"
        >
          by TestUser
        </span>
      </div>
      <div
        className="body undefined"
      >
        This is not a real team. You have been deceived...
      </div>
      <div />
    </a>
  </div>
</div>
`;
